src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,639 @@
+     * @param t the service iterator
+     * @param algorithm the algorithm
+     * @param kdfParameters the algorithm parameters
+     */
+    private KDF(KDFSpi firstSpi, Provider firstProv, Iterator<Service> t, String algorithm,
+                KDFParameters kdfParameters) {
+        this.firstSpi = firstSpi;
+        this.firstProv = firstProv;
+        serviceIterator = t;
+        this.algorithm = algorithm;
+        this.kdfParameters = kdfParameters;
+        lock = new Object();
+    }
+
+    /**
+     * Returns the algorithm name of this {@code KDF} object.
+     *
+     * @return the algorithm name of this {@code KDF} object
+     */
+    public String getAlgorithm() {
+        return this.algorithm;
+    }
+
+    /**
+     * Returns the name of the provider.
+     *
+     * @return the name of the provider
+     */
+    public String getProviderName() {
+        useFirstSpi();
+        return provider.getName();
+    }
+
+    /**
+     * Returns the {@code KDFParameters} used with this {@code KDF} object.
+     * <p>
+     * The returned parameters may be the same that were used to initialize
+     * this {@code KDF} object, or may contain additional default or
+     * random parameter values used by the underlying KDF algorithm.
+     * If the required parameters were not supplied and can be generated by
+     * the {@code KDF} object, the generated parameters are returned;
+     * otherwise {@code null} is returned.
+     *
+     * @return the parameters used with this {@code KDF} object, or
+     * {@code null}
+     */
+    public KDFParameters getParameters() {
+        useFirstSpi();
+        return spi.engineGetParameters();
+    }
+
+    /**
+     * Returns a {@code KDF} instance initialized with the specified algorithm.
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use.
+     *     See the {@code KDF} section in the <a href=
+     *     "{@docRoot}/../specs/security/standard-names.html#kdf-algorithms">
+     *     Java Security Standard Algorithm Names Specification</a>
+     *     for information about standard KDF algorithm names.
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if no {@code Provider} supports a {@code KDF} implementation for the
+     *     specified algorithm
+     * @throws NullPointerException
+     *     if {@code algorithm} is {@code null}
+     */
+    public static KDF getInstance(String algorithm)
+        throws NoSuchAlgorithmException {
+        Objects.requireNonNull(algorithm, "algorithm should not be null");
+        try {
+            return getInstance(algorithm, (KDFParameters) null);
+        } catch (InvalidAlgorithmParameterException e) {
+            throw new NoSuchAlgorithmException(
+                "No implementation found using null KDFParameters", e);
+        }
+    }
+
+    /**
+     * Returns a {@code KDF} instance initialized with the specified algorithm from
```

Comment:
"initialized with the specified algorithm" is not really accurate. I would change this text to "that implements the specified algorithm". Applies to all `getInstance` methods.

Reply:
I find Sean's suggestion more accurate than the "initialized" word - even if we have not selected the provider, the final choice should "implement" the specified algorithm.